### PR TITLE
construction: treat already-in-mempool as success

### LINKED ISSUE #39: make construction submit succeed if tx is already in mempool
> maybe we should hook up "transaction already in mempool" to return success

Yeah this would probably be in line with Rosetta semantics (no error if transaction is added to mempool). We could add a proper error type for this on the Oasis Core side so that you wouldn't need to match strings (we support transporting actual error codes across RPC boundaries).

_Originally posted by @kostko in https://github.com/oasisprotocol/oasis-core-rosetta-gateway/pull/36#issuecomment-673303278_

### PR DESCRIPTION
**Pull request by pro-wh**
fixes #39 

:sparkles: `errors.Is` :sparkles:

### CODE REVIEW

File: services/construction.go [hunk 1]
```diff
@@ -127,7 +129,12 @@ func (s *constructionAPIService) ConstructionSubmit(
 
 	if err := s.oasisClient.SubmitTxNoWait(ctx, request.SignedTransaction); err != nil {
 		loggerCons.Error("ConstructionSubmit: SubmitTxNoWait failed", "err", err)
-		return nil, ErrUnableToSubmitTx
+		// TODO: Have oasis-core use a distinct error code for this.
```
Comment (kostko) [inline services/construction.go:132]: Please file an issue, it should be trivial to support a proper error code.

File: services/construction.go [hunk 2]
```diff
@@ -17,6 +18,7 @@ import (
 	"github.com/oasisprotocol/oasis-core/go/common/quantity"
 	"github.com/oasisprotocol/oasis-core/go/consensus/api/transaction"
 	staking "github.com/oasisprotocol/oasis-core/go/staking/api"
+	"github.com/tendermint/tendermint/mempool"
```
Comment (kostko) [inline services/construction.go:21]: This sounds like a very bad idea, this should be done in Oasis Core and exposed as a `ErrDuplicateTx`, not exposing internal details. I would rather you just hardcode the error string here and add an `XXX` comment below, as the used version of Tendermint (added to `go.mod` above) is inconsistent with what Oasis Core is actually using anyway.

File: services/construction.go (hunk 1 above)
Comment (kostko) [reply services/construction.go:132]: Filed as https://github.com/oasisprotocol/oasis-core/issues/3192 implemented in https://github.com/oasisprotocol/oasis-core/pull/3193.

File: services/construction.go (hunk 2 above)
Comment (pro-wh) [reply services/construction.go:21]: yeah, it is uncool to add such a big module as a direct dependency for a constant value

let's just wait for another release of oasis-core. whether this call succeeds or fails isn't affecting anything by much. it's still the rosetta client's responsibility to rebroadcast once in a while even if it's indicated that it succeeds

**Comment by pro-wh** [comment]
preparing this for a new oasis-core release with the dedicated error

File: services/construction.go (hunk 2 above)
Comment (pro-wh) [reply services/construction.go]: we have now waited for that

File: services/construction.go (hunk 1 above)
Comment (pro-wh) [reply services/construction.go]: changed it to that

**Comment by pro-wh** [comment]
the time has come! this is updated to use oasis-core 20.9.2, with the new dedicated `ErrDuplicateTx`

**Review by peterjgilbert — APPROVED** [review]

**Pull request merged**